pkg/steps/source.go

```diff
@@ -37,22 +38,40 @@ const (
 
 	ProwJobIdLabel = "prow.k8s.io/id"
 
-	gopath = "/go"
+	gopath        = "/go"
+	sshPrivateKey = "/sshprivatekey"
+	sshConfig     = "/ssh_config"
 )
 
 var (
 	JobSpecAnnotation = fmt.Sprintf("%s/%s", CiAnnotationPrefix, "job-spec")
 )
 
-func sourceDockerfile(fromTag api.PipelineImageStreamTagReference, workingDir string) string {
-	return fmt.Sprintf(`
-FROM %s:%s
-ADD ./app.binary /clonerefs
-RUN umask 0002 && /clonerefs && find %s/src -type d -not -perm -0775 | xargs -r chmod g+xw
-WORKDIR %s/
-ENV GOPATH=%s
-RUN git submodule update --init
-`, api.PipelineImageStream, fromTag, gopath, workingDir, gopath)
+func sourceDockerfile(fromTag api.PipelineImageStreamTagReference, workingDir string, sshSecretName string) string {
+	var dockerCommands []string
+	dockerCommands = append(dockerCommands, "")
+	dockerCommands = append(dockerCommands, fmt.Sprintf("FROM %s:%s", api.PipelineImageStream, fromTag))
+	dockerCommands = append(dockerCommands, "ADD ./app.binary /clonerefs")
+
+	if len(sshSecretName) > 0 {
+		dockerCommands = append(dockerCommands, fmt.Sprintf("ADD %s /etc/ssh/ssh_config", sshConfig))
+		dockerCommands = append(dockerCommands, fmt.Sprintf("COPY ./%s %s", coreapi.SSHAuthPrivateKey, sshPrivateKey))
+	}
+
+	dockerCommands = append(dockerCommands, fmt.Sprintf("RUN umask 0002 && /clonerefs && find %s/src -type d -not -perm -0775 | xargs -r chmod g+xw", gopath))
+	dockerCommands = append(dockerCommands, fmt.Sprintf("WORKDIR %s/", workingDir))
+	dockerCommands = append(dockerCommands, fmt.Sprintf("ENV GOPATH=%s", gopath))
+	dockerCommands = append(dockerCommands, "RUN git submodule update --init")
+
+	// After the clonerefs command, we don't need the secret anymore.
+	// We don't want to let the key keep existing in the image's layer.
```

Comment:
Isn't the key present in an intermediate layer though? If so, is it a problem (are we pushing that image somewhere publicly available)?

Reply:
All the build inputs are deleted and are not are available while building. Therefore we need to `ADD` the key somewhere to be used from `clonerefs`. If we don't delete it, the keys will be available in the image layer and anyone can access it if the image will be pulled.

Reply:
> Users should normally remove their input secrets from the final application image so that the secrets are not present in the container running from that image. However, the secrets will still exist in the image itself in the layer where they were added. This removal should be part of the Dockerfile itself.

ref: https://docs.openshift.com/container-platform/3.5/dev_guide/builds/build_inputs.html#using-secrets-docker-strategy

Reply:
I'm not saying it should not be deleted. I'm wondering if deleting it from the outermost layer is actually enough to protect it: the bits are still present in an intermediate layer. Unless I'm missing some OpenShift build smartness, but this was a gotcha with Docker.

Reply:
The initial secret will be injecting as a build input. After the build, the bytes won't be available. I guess this is some Openshift builds magic....

Reply:
I want to to test this a bit: this is sensitive area, let's be 100% sure. Also @bbguimaraes mentioned in Slack something that supports this should be fine.

Reply:
Let's ask DevEx team on this please